<h1>Fórmula 1 + Engenharia de Dados?</h1>

<p>Nos inspiramos no melhor técnico de pipelines (encanador), juntamos com corrida e nomeamos nossa equipe de ...</p>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<h1 style="text-align: center;">Equipe Mario Kart</h1>

!["Equipe Mario Kart"](./images/mario_kart1.jpg)

<h1 style="text-align: center;">Será dada a largada, acendam-se as luzes vermelhas...</h1>

!["Equipe Mario Kart"](./images/f1_start.jpg)

<h3>Regras de Negócios e Questões a serem respondidas</h3>

<p>A FIA (Federação Internacional de Automobilismo) deseja construir um banco de dados sobre o campeonato de Fórmula 1. Para isso contratou uma consultoria de software e você foi alocado no projeto como analista de sistemas responsável pelo projeto conceitual do banco de dados.</p>
    
<p>Após um levantamento de requisitos você identificou as informações relevantes abaixo:</p>

<ul>
    <li>Os contrutores contratam pilotos</li>
    <li>Os construtores contratam fabricantes de motores</li>
    <li>Os circuitos recebem os grandes prêmios</li>
</ul>

<p><strong>Os grandes prêmios possuem a seguinte pontuação:</strong></p>
    
<table>
    <tr>
        <td style="text-align: center;">1º</td>
        <td style="text-align: center;">2º</td>
        <td style="text-align: center;">3º</td>
        <td style="text-align: center;">4º</td>
        <td style="text-align: center;">5º</td>
        <td style="text-align: center;">6º</td>
        <td style="text-align: center;">7º</td>
        <td style="text-align: center;">8º</td>
        <td style="text-align: center;">9º</td>
        <td style="text-align: center;">10º</td>
    </tr>
    <tr>
        <td style="text-align: center;">25</td>
        <td style="text-align: center;">18</td>
        <td style="text-align: center;">15</td>
        <td style="text-align: center;">12</td>
        <td style="text-align: center;">10</td>
        <td style="text-align: center;">8</td>
        <td style="text-align: center;">6</td>
        <td style="text-align: center;">4</td>
        <td style="text-align: center;">2</td>
        <td style="text-align: center;">1</td>
    </tr>
</table>

<p><strong>Observação:</strong></p>

<ul>
    <li>Caso o piloto faça a volta mais rápida e termine entre os 10 primeiros, o mesmo ganha um ponto adicional.</li>
    <li>O pódio é formado pelos 3 primeiros colocados.</li>
    <li>Cada grande prêmio possui um treino classificatório, que decide o Grid de largada.</li>
</ul>

<p><strong>Os requisitos de informação solicitados pelos usuários foram:</strong></p>

<ul>
    <li>Classificação do campeonado de pilotos</li>
    <li>Classificação do campeonado de contrutores</li>
    <li>Pilotos com mais pódios</li>
    <li>Pilotos com mais poles (primeiro lugar no treino classificatório)</li>
    <li>Após a modelagem do banco de dados relacional, você deve modelar um Data Warehouse para este projeto.</li>
    <li>Baseando no Data Warehouse, faça explorações de dados com Python, gerando insights sobre a massa de dados à sua escolha.</li>
</ul>

<h2>Entendimento dos dados:</h2>

<p>Utilizamos a base de dados disponibilizada no <a href="https://www.kaggle.com/rohanrao/formula-1-world-championship-1950-2020?select=results.csv" target="_blank">Kaggle</a>.</p> 

<p>Inclusive agradecimentos especiais ao Rohan Rao, Cientista de Dados, Indiano, que disponibilizou os datasets <a href="https://www.kaggle.com/rohanrao" target="_blank">Link com dados dele</a>. O Rohan é Quadruple Kaggle Grandmaster</p>

<p>Após analisar os datasets e entendê-los, passamos para a fase da modelagem conceitual do nosso projeto</p>

<h2 style="text-align: center">Modelo conceitual</h2>

!["Modelo Conceitual - Equipe Mario Kart"](./images/modelo_conceitual.jpg)

<h3>Entendimento do modelo conceitual</h3>

<p>Seguindo os conceitos de banco de dados relacional, buscamos normalizar as tabelas para evitar redundância.</p>

<p>Um campeonato de Fórmula 1 é composto de algumas entidades e através dessas entidades que modelamos o coneitual do projeto. 
O construtor tem pilotos, fabricante de motor, origem de um país e também resultados das corridas. 
Um Construtor pode ter um ou mais pilotos e um fabricante de motor, enquanto um fabricante de motor pode fornecer motor para um ou mais construtores.</p>

<p>A entidade Fabricante de Motor se relaciona com a entidade construtor. Fabricante de motor fornece motor para um ou mais construtor.</p>

<p>O Piloto relaciona-se com as entidades construtor, país, resultado das corridas e pit stops, onde o piloto tem um construtor e um país de origem, pode ter um ou mais resultados das corridas e um ou mais pit stops realizados nas corridas.</p>

<p>A Entidade País relaciona-se com as entidades construtor, piloto e circuito. Construtor e piloto tem um país de origem, o circuito tem um país onde é realizado.</p>

<p>A entidade Circuito se relaciona com as entidades corrida e país. O circuito tem uma ou mais corridas e um país onde está sendo realizado.</p>

<p>Entidade Corrida relaciona-se com as entidades circuito, resultado das corridas e pit stops. Uma corrida possui um circuito, um ou mais pit stops e um ou mais resultados.</p>

<p>Já a Entidade Pit Stop se relaciona com as entidades piloto e corrida. Onde um pit stop é feito por um piloto em uma determinada corrida.</p>

<p>A entidade Resultado relaciona-se com as entidades construtor, piloto, corrida e status. Um resultado é de um piloto, de um construtor, de uma determinada corrida e com um status.</p>

<p>A entidade Status se relaciona com a entidade Resultado. Onde um status é de um resultado.</p>


<h2 style="text-align: center">Modelo físico</h2>

!["Modelo Conceitual - Equipe Mario Kart"](./images/modelo_fisico.jpg)

<h3>Código gerado para criação do banco de dados relacional</h3>

``` sql
CREATE DATABASE DB_MARIO_KART;

USE DB_MARIO_KART;

/*==============================================================*/
/* Table: TB_PAIS                                               */
/*==============================================================*/
CREATE TABLE TB_PAIS (
   ID_PAIS INTEGER NOT NULL auto_increment, 
   NOME_PAIS VARCHAR(40) NOT NULL UNIQUE,
   CONSTRAINT PK_TB_PAIS PRIMARY KEY (ID_PAIS)
);

/*==============================================================*/
/* Table: TB_FABRICANTE_MOTOR                                   */
/*==============================================================*/
CREATE TABLE TB_FABRICANTE_MOTOR (
   ID_FABRICANTE INTEGER NOT NULL auto_increment,
   NOME_FABRICANTE VARCHAR(40) NOT NULL,
   CONSTRAINT PK_TB_FABRICANTE_MOTOR PRIMARY KEY (ID_FABRICANTE)
);

/*==============================================================*/
/* Table: TB_STATUS                                             */
/*==============================================================*/
CREATE TABLE TB_STATUS (
   ID_STATUS  INTEGER NOT NULL,
   DESCRICAO  VARCHAR(40),
   CONSTRAINT PK_TB_STATUS PRIMARY KEY (ID_STATUS)
);

/*==============================================================*/
/* Table: TB_CONSTRUTOR                                         */
/*==============================================================*/
CREATE TABLE TB_CONSTRUTOR (
   ID_CONSTRUTOR INTEGER  NOT NULL,
   ID_PAIS INTEGER NOT NULL,
   ID_FABRICANTE INTEGER NOT NULL,
   NOME_CONSTRUTOR VARCHAR(40) NOT NULL,
   CONSTRAINT PK_TB_CONSTRUTOR PRIMARY KEY (ID_CONSTRUTOR),
   CONSTRAINT FK_TB_CONSTRUTOR_TB_PAIS FOREIGN KEY (ID_PAIS) REFERENCES TB_PAIS (ID_PAIS),
   CONSTRAINT FK_TB_CONSTRUTOR_TB_FABRICANTE_MOTOR FOREIGN KEY (ID_FABRICANTE) REFERENCES TB_FABRICANTE_MOTOR (ID_FABRICANTE)   
);

/*==============================================================*/
/* Table: TB_PILOTO                                             */
/*==============================================================*/
CREATE TABLE TB_PILOTO (
   ID_PILOTO            INTEGER               NOT NULL,
   ID_PAIS              INTEGER               NOT NULL,
   ID_CONSTRUTOR        INTEGER               NOT NULL,
   REF_PILOTO           VARCHAR(25)           NOT NULL,
   SIGLA_PILOTO         VARCHAR(3)            NOT NULL,
   NOME_PILOTO          VARCHAR(40)           NOT NULL,
   SOBRENOME_PILOTO     VARCHAR(40)           NOT NULL,
   DATA_NASCIMENTO      DATE,
   CONSTRAINT PK_TB_PILOTO PRIMARY KEY (ID_PILOTO),
   CONSTRAINT FK_TB_PILOTO_TB_PAIS FOREIGN KEY (ID_PAIS) REFERENCES TB_PAIS (ID_PAIS),
   CONSTRAINT FK_TB_PILOTO_TB_CONSTRUTOR FOREIGN KEY (ID_CONSTRUTOR) REFERENCES TB_CONSTRUTOR (ID_CONSTRUTOR) 
);

/*==============================================================*/
/* Table: TB_CIRCUITO                                           */
/*==============================================================*/

CREATE TABLE TB_CIRCUITO (
   ID_CIRCUITO          INTEGER					 NOT NULL,
   ID_PAIS              INTEGER					 NOT NULL,
   LOCALIDADE           VARCHAR(40)              NOT NULL,
   NOME                 VARCHAR(40)              NOT NULL,
   LATITUDE             FLOAT,
   LONGITUDE            FLOAT,
   ALTITUDE             INTEGER,
   CONSTRAINT PK_TB_CIRCUITO PRIMARY KEY (ID_CIRCUITO),
   CONSTRAINT FK_TB_CIRCUITO_TB_PAIS FOREIGN KEY (ID_PAIS) REFERENCES TB_PAIS (ID_PAIS)
);

/*==============================================================*/
/* Table: TB_CORRIDA                                            */
/*==============================================================*/
CREATE TABLE TB_CORRIDA (
   ID_CORRIDA             INTEGER               NOT NULL,
   ID_CIRCUITO            INTEGER               NOT NULL,
   ANO                    INTEGER               NOT NULL,
   NOME                   VARCHAR(40)           NOT NULL,
   ORDEM_CORRIDA          INTEGER,
   CONSTRAINT PK_TB_CORRIDA PRIMARY KEY (ID_CORRIDA),
   CONSTRAINT FK_TB_CORRIDA_TB_CIRCUITO FOREIGN KEY (ID_CIRCUITO) REFERENCES TB_CIRCUITO (ID_CIRCUITO)
);

/*==============================================================*/
/* Table: TB_PIT_STOPS                                          */
/*==============================================================*/
CREATE TABLE TB_PIT_STOPS (
   ID_PIT_STOP          INTEGER               NOT NULL          auto_increment,
   ID_CORRIDA           INTEGER               NOT NULL,
   ID_PILOTO            INTEGER               NOT NULL,
   TEMPO_PIT_STOP       INTEGER               NOT NULL,
   NO_VOLTA             INTEGER               NOT NULL,
   CONSTRAINT PK_TB_PIT_STOPS PRIMARY KEY (ID_PIT_STOP),
   CONSTRAINT FK_TB_PIT_STOPS_TB_CORRIDA FOREIGN KEY (ID_CORRIDA) REFERENCES TB_CORRIDA (ID_CORRIDA),
   CONSTRAINT FK_TB_PIT_STOPS_TB_PILOTO FOREIGN KEY (ID_PILOTO) REFERENCES TB_PILOTO (ID_PILOTO) 
);

/*==============================================================*/
/* Table: TB_RESULTADO                                          */
/*==============================================================*/
CREATE TABLE TB_RESULTADO (
   ID_RESULTADO         INTEGER               NOT NULL,
   ID_CORRIDA           INTEGER               NOT NULL,
   ID_PILOTO            INTEGER               NOT NULL,
   ID_CONSTRUTOR        INTEGER               NOT NULL,
   ID_STATUS            INTEGER               NOT NULL,
   POS_LARGADA          INTEGER               NOT NULL,
   POS_FINAL            INTEGER               NOT NULL,
   PONTOS_CORRIDA       INTEGER,
   RANK_VOLTA_RAPIDA    INTEGER,
   PONTO_VOLTA_RAPIDA   INTEGER               NOT NULL,
   CONSTRAINT PK_TB_RESULTADO PRIMARY KEY (ID_RESULTADO),
   CONSTRAINT FK_TB_RESULTADO_TB_CORRIDA FOREIGN KEY (ID_CORRIDA) REFERENCES TB_CORRIDA (ID_CORRIDA),
   CONSTRAINT FK_TB_RESULTADO_TB_PILOTO FOREIGN KEY (ID_PILOTO) REFERENCES TB_PILOTO (ID_PILOTO),
   CONSTRAINT FK_TB_RESULTADO_TB_CONSTRUTOR FOREIGN KEY (ID_CONSTRUTOR) REFERENCES TB_CONSTRUTOR (ID_CONSTRUTOR), 
   CONSTRAINT FK_TB_RESULTADO_TB_STATUS FOREIGN KEY (ID_STATUS) REFERENCES TB_STATUS (ID_STATUS)
);

/*==============================================================*/
/* Index: TB_PIT_STOPS                                    */
/*==============================================================*/
CREATE INDEX IDX_TB_PITSTOPS ON TB_PIT_STOPS (ID_PILOTO, ID_CORRIDA);

/*==============================================================*/
/* Index: TB_RESULTADO                                    */
/*==============================================================*/
CREATE INDEX IDX_TB_RESULTADO ON TB_RESULTADO (ID_CORRIDA, ID_PILOTO, ID_CONSTRUTOR, ID_STATUS);



<h2 style="text-align: center">Modelo Dimensional</h2>

!["Modelo DW - Equipe Mario Kart"](./images/modelo_dimensional.jpg)

<h3>Código gerado para criação do banco de dados relacional</h3>

``` sql
CREATE DATABASE DW_MARIO_KART;

USE DW_MARIO_KART;

/*==============================================================*/
/* Table: DIM_CONSTRUTOR                                        */
/*==============================================================*/
CREATE TABLE DIM_CONSTRUTOR (
   ID_CONSTRUTOR        INTEGER              NOT NULL,
   PAIS                 VARCHAR(40)          NOT NULL,
   FABRICANTE_MOTOR     VARCHAR(40)          NOT NULL,
   NOME_CONSTRUTOR      VARCHAR(40)          NOT NULL,
   CONSTRAINT PK_DIM_CONSTRUTOR PRIMARY KEY (ID_CONSTRUTOR)
);

/*==============================================================*/
/* Table: DIM_CORRIDA                                           */
/*==============================================================*/
CREATE TABLE DIM_CORRIDA (
   ID_CORRIDA           INTEGER               NOT NULL,
   NOME_CORRIDA         VARCHAR(40)           NOT NULL,
   NOME_CIRCUITO        VARCHAR(40)           NOT NULL,
   ANO                  INTEGER               NOT NULL,
   ORDEM_CORRIDA        INTEGER,
   PAIS                 VARCHAR(40)           NOT NULL,
   CONSTRAINT PK_DIM_CORRIDA PRIMARY KEY (ID_CORRIDA)
);

/*==============================================================*/
/* Table: DIM_PILOTO                                            */
/*==============================================================*/
CREATE TABLE DIM_PILOTO (
   ID_PILOTO            INTEGER              NOT NULL,
   PAIS                 VARCHAR(40)          NOT NULL,
   NOME_CONSTRUTOR      VARCHAR(40)          NOT NULL,
   REF_PILOTO           VARCHAR(25)          NOT NULL,
   SIGLA_PILOTO         VARCHAR(3)           NOT NULL,
   NOME_PILOTO          VARCHAR(40)          NOT NULL,
   SOBRENOME_PILOTO     VARCHAR(40)          NOT NULL,
   DATA_NASCIMENTO      DATE,
   CONSTRAINT PK_DIM_PILOTO PRIMARY KEY (ID_PILOTO)
);

/*==============================================================*/
/* Table: FATO_RESULTADO                                        */
/*==============================================================*/
CREATE TABLE FATO_RESULTADO (
   ID_RESULTADO         INTEGER              NOT NULL,
   ID_CORRIDA           INTEGER              NOT NULL,
   ID_PILOTO            INTEGER              NOT NULL,
   ID_CONSTRUTOR        INTEGER              NOT NULL,
   POS_LARGADA          INTEGER              NOT NULL,
   POS_FINAL            INTEGER              NOT NULL,
   PONTOS_CORRIDA       INTEGER,
   RANK_VOLTA_RAPIDA    INTEGER,
   PONTO_VOLTA_RAPIDA   INTEGER,
   STATUS               VARCHAR(40)          NOT NULL,
   CONSTRAINT PK_FATO_RESULTADO PRIMARY KEY (ID_RESULTADO),
   CONSTRAINT FK_FATO_RESULTADO_DIM_CORRIDA FOREIGN KEY (ID_CORRIDA) REFERENCES DIM_CORRIDA (ID_CORRIDA),
   CONSTRAINT FK_FATO_RESULTADO_DIM_PILOTO FOREIGN KEY (ID_PILOTO) REFERENCES DIM_PILOTO (ID_PILOTO), 
   CONSTRAINT FK_FATO_RAULTADO_DIM_CONSTRUTOR FOREIGN KEY (ID_CONSTRUTOR) REFERENCES DIM_CONSTRUTOR (ID_CONSTRUTOR)
);


```


<h3>Script para popular o DW a partir do Relacional</h3>

```sql
/*==============================================================*/
/* Table: DIM_CONSTRUTOR                                        */
/*==============================================================*/
INSERT INTO DIM_CONSTRUTOR(ID_CONSTRUTOR, PAIS, FABRICANTE_MOTOR, NOME_CONSTRUTOR)
SELECT construtor.ID_CONSTRUTOR, pais.NOME_PAIS, fabricante.NOME_FABRICANTE, construtor.NOME_CONSTRUTOR
FROM db_mario_kart.tb_construtor construtor
	JOIN db_mario_kart.tb_pais pais ON construtor.ID_PAIS = pais.ID_PAIS
	JOIN db_mario_kart.tb_fabricante_motor fabricante ON construtor.ID_FABRICANTE = fabricante.ID_FABRICANTE;
    

/*==============================================================*/
/* Table: DIM_CORRIDA                                        */
/*==============================================================*/
INSERT INTO DIM_CORRIDA(ID_CORRIDA, NOME_CORRIDA, NOME_CIRCUITO, ANO, ORDEM_CORRIDA, PAIS)
SELECT corrida.ID_CORRIDA, corrida.NOME, circuito.NOME, corrida.ANO, corrida.ORDEM_CORRIDA, pais.NOME_PAIS
FROM db_mario_kart.tb_corrida corrida
	JOIN db_mario_kart.tb_circuito circuito ON corrida.ID_CIRCUITO = circuito.ID_CIRCUITO
	JOIN db_mario_kart.tb_pais pais ON circuito.ID_PAIS = pais.ID_PAIS;
    

/*==============================================================*/
/* Table: DIM_PILOTO                                        */
/*==============================================================*/
INSERT INTO DIM_PILOTO(ID_PILOTO, PAIS, NOME_CONSTRUTOR, REF_PILOTO, SIGLA_PILOTO, NOME_PILOTO, SOBRENOME_PILOTO, DATA_NASCIMENTO)
SELECT piloto.ID_PILOTO, pais.NOME_PAIS, construtor.NOME_CONSTRUTOR, piloto.REF_PILOTO, piloto.SIGLA_PILOTO, piloto.NOME_PILOTO, piloto.SOBRENOME_PILOTO, piloto.DATA_NASCIMENTO
FROM db_mario_kart.tb_piloto piloto
	JOIN db_mario_kart.tb_pais pais ON piloto.ID_PAIS = pais.ID_PAIS
    JOIN db_mario_kart.tb_construtor construtor ON piloto.ID_CONSTRUTOR = construtor.ID_CONSTRUTOR;
    

/*==============================================================*/
/* Table: FATO_RESULTADO                                        */
/*==============================================================*/
INSERT INTO FATO_RESULTADO(ID_RESULTADO, ID_CORRIDA, ID_PILOTO, ID_CONSTRUTOR, POS_LARGADA, POS_FINAL, PONTOS_CORRIDA, RANK_VOLTA_RAPIDA, PONTO_VOLTA_RAPIDA, STATUS)
SELECT resultado.ID_RESULTADO, resultado.ID_CORRIDA, resultado.ID_PILOTO, resultado.ID_CONSTRUTOR, resultado.POS_LARGADA, resultado.POS_FINAL, resultado.PONTOS_CORRIDA, resultado.RANK_VOLTA_RAPIDA, resultado.PONTO_VOLTA_RAPIDA, status.DESCRICAO
FROM db_mario_kart.tb_resultado resultado
	JOIN db_mario_kart.tb_status status ON resultado.ID_STATUS = status.ID_STATUS;


```

<h3>Vamos agora conectar ao nosso banco dimensional</h3>

<p>Além disso, vamos salvar as tabelas no Dataframe do Pandas para facilitar nosso trabalho!</p>

In [1]:
# Importando o pandas e o sqlalchemy para conseguir conectar ao banco
import pandas as pd
import sqlalchemy

In [2]:
# Instalando a biblioteca pymsql para conectar ao banco mysql
!pip install pymysql

In [3]:
# Criando a engine de conexão
engine = sqlalchemy.create_engine('mysql+pymysql://gama:gama@127.0.0.1:3306/dw_mario_kart')

In [163]:
# Criando um DataFrame a partir de uma query utilizando parâmetros dinâmicos.

query = 'SELECT *' \
      'FROM dim_construtor ' \
      'WHERE nome_construtor = %s'
df3 = pd.read_sql_query(query, engine,params=[input('digite um construtor: \n')])
df3.head()

digite um construtor: 
McLaren


ID_CONSTRUTOR            PAIS FABRICANTE_MOTOR NOME_CONSTRUTOR
0              1  United Kingdom         Mercedes         McLaren

### Vamos requisitar as tabelas do DW que vamos precisar consultar e salvar num dataframe do pandas

In [6]:
# Salvando as tabelas das tabelas do DW em dataframes do pandas
fato_resultado = pd.read_sql_table('fato_resultado', engine)
dim_construtor = pd.read_sql_table('dim_construtor', engine)
dim_corrida = pd.read_sql_table('dim_corrida', engine)
dim_piloto = pd.read_sql_table('dim_piloto', engine)

#### Vamos verificar os dados e o shape dos dataframes se confere com a origem

In [7]:
fato_resultado.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  POS_FINAL  \
0             1          18          1              1            1          1   
1             2          18          2              2            5          2   
2             3          18          3              3            7          3   
3             4          18          4              4           11          4   
4             5          18          5              1            3          5   

   PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA    STATUS  
0              10                  2                   0  Finished  
1               8                  3                   0  Finished  
2               6                  5                   0  Finished  
3               5                  7                   0  Finished  
4               4                  1                   1  Finished

In [71]:
# Quantas linhas e colunas no dataframe fato_resultado
fato_resultado.shape

(25040, 10)

In [72]:
# Não existe valores nulos
fato_resultado.isnull().sum()

ID_RESULTADO          0
ID_CORRIDA            0
ID_PILOTO             0
ID_CONSTRUTOR         0
POS_LARGADA           0
POS_FINAL             0
PONTOS_CORRIDA        0
RANK_VOLTA_RAPIDA     0
PONTO_VOLTA_RAPIDA    0
STATUS                0
dtype: int64

In [73]:
fato_resultado.describe()

# Encontramos um valor máximo para pontos_corrida de 50 pontos, será que isso está correto?

ID_RESULTADO    ID_CORRIDA     ID_PILOTO  ID_CONSTRUTOR   POS_LARGADA  \
count  25040.000000  25040.000000  25040.000000   25040.000000  25040.000000   
mean   12521.201238    514.052037    247.809944      47.130831     11.215375   
std     7229.580124    287.705967    255.124887      57.977798      7.281000   
min        1.000000      1.000000      1.000000       1.000000      0.000000   
25%     6260.750000    285.750000     56.000000       6.000000      5.000000   
50%    12520.500000    499.000000    158.000000      25.000000     11.000000   
75%    18780.250000    756.000000    346.000000      57.000000     17.000000   
max    25045.000000   1055.000000    854.000000     214.000000     34.000000   

          POS_FINAL  PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA  
count  25040.000000    25040.000000       25040.000000        25040.000000  
mean      12.951917        1.776917           2.850200            0.011941  
std        7.754542        3.995294           5.673066            0.108622  
min        1.000000        0.000000           0.000000            0.000000  
25%        6.000000        0.000000           0.000000            0.000000  
50%       12.000000        0.000000           0.000000            0.000000  
75%       19.000000        2.000000           2.000000            0.000000  
max       39.000000       50.000000          24.000000            1.000000

In [74]:
fato_resultado[fato_resultado['PONTOS_CORRIDA'] > 26]

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  \
22514         22518         918          1            131            2   
22515         22519         918         13              3            4   
22516         22520         918        822              3            3   

       POS_FINAL  PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA  \
22514          1              50                  4                   0   
22515          2              36                  2                   0   
22516          3              30                  6                   0   

         STATUS  
22514  Finished  
22515  Finished  
22516  Finished

<p>Verificamos que a corrida de ID 918 foi a única com uma pontuação diferente, com valores maiores de 26 pontos (que seria 25 pontos do primeiro lugar + 1 ponto de volta rápida)</p>

In [75]:
dim_corrida[dim_corrida['ID_CORRIDA'] == 918]

ID_CORRIDA          NOME_CORRIDA       NOME_CIRCUITO   ANO  \
915         918  Abu Dhabi Grand Prix  Yas Marina Circuit  2014   

     ORDEM_CORRIDA                  PAIS  
915             19  United Arab Emirates

<h3>Alterações na estrutura de pontos</h3>
<p>O Grande Prêmio de Abu Dhabi de 2014 foi uma corrida de Fórmula 1 realizada no Circuito de Yas Marina em 23 de novembro de 2014. A corrida foi a décima nona e última etapa da temporada de 2014.
    
Pela primeira vez na história da Fórmula 1, equipes e pilotos marcaram o dobro do número de pontos conquistados para as posições de chegada da corrida. A FIA implementou isso para maximizar o foco no campeonato até o final da temporada. Originalmente Bernie Ecclestone queria pontos duplos para as últimas três corridas da temporada,mas as equipes finalmente decidiram ter pontos duplos apenas para a última corrida da temporada. A mudança de regra foi recebida negativamente por equipes e pilotos. <a href="https://en.wikipedia.org/wiki/2014_Abu_Dhabi_Grand_Prix" target="_blank">Link da informação</a></p>

In [76]:
# Analisando os dados da dimensao construtor
dim_construtor.head()

ID_CONSTRUTOR            PAIS FABRICANTE_MOTOR NOME_CONSTRUTOR
0              1  United Kingdom         Mercedes         McLaren
1              2         Germany              BMW      BMW Sauber
2              3  United Kingdom         Mercedes        Williams
3              4          France          Renault         Renault
4              5           Italy            Honda      Toro Rosso

In [77]:
# Quantidade de linhas e colunas da dimensao construtor
dim_construtor.shape

(211, 4)

In [79]:
# Sem valores nulos também
dim_construtor.isnull().sum()

ID_CONSTRUTOR       0
PAIS                0
FABRICANTE_MOTOR    0
NOME_CONSTRUTOR     0
dtype: int64

In [80]:
# Analisando os dados da dimensão corrida
dim_corrida.head()

ID_CORRIDA           NOME_CORRIDA                   NOME_CIRCUITO   ANO  \
0           1  Australian Grand Prix  Albert Park Grand Prix Circuit  2009   
1           2   Malaysian Grand Prix    Sepang International Circuit  2009   
2           3     Chinese Grand Prix  Shanghai International Circuit  2009   
3           4     Bahrain Grand Prix   Bahrain International Circuit  2009   
4           5     Spanish Grand Prix  Circuit de Barcelona-Catalunya  2009   

   ORDEM_CORRIDA       PAIS  
0              1  Australia  
1              2   Malaysia  
2              3      China  
3              4    Bahrain  
4              5      Spain

In [81]:
# Quantidade de linhas e colunas da dimensão corrida
dim_corrida.shape

(1058, 6)

In [82]:
# Sem valores nulos na dimensao corrida
dim_corrida.isnull().sum()

ID_CORRIDA       0
NOME_CORRIDA     0
NOME_CIRCUITO    0
ANO              0
ORDEM_CORRIDA    0
PAIS             0
dtype: int64

In [83]:
# Analisando os dados da dimensâo piloto
dim_piloto.head()

ID_PILOTO            PAIS NOME_CONSTRUTOR  REF_PILOTO SIGLA_PILOTO  \
0          1  United Kingdom         McLaren    hamilton          HAM   
1          2         Germany      BMW Sauber    heidfeld          HEI   
2          3         Germany        Williams     rosberg          ROS   
3          4           Spain         Renault      alonso          ALO   
4          5         Finland         McLaren  kovalainen          KOV   

  NOME_PILOTO SOBRENOME_PILOTO DATA_NASCIMENTO  
0       Lewis         Hamilton      1985-01-07  
1        Nick         Heidfeld      1977-05-10  
2        Nico          Rosberg      1985-06-27  
3    Fernando           Alonso      1981-07-29  
4      Heikki       Kovalainen      1981-10-19

In [84]:
# Quantidade de linhas e colunas da dimensao piloto
dim_piloto.shape

(853, 8)

In [85]:
# Sem valores nulos na dimensão piloto
dim_piloto.isnull().sum()

ID_PILOTO           0
PAIS                0
NOME_CONSTRUTOR     0
REF_PILOTO          0
SIGLA_PILOTO        0
NOME_PILOTO         0
SOBRENOME_PILOTO    0
DATA_NASCIMENTO     0
dtype: int64

#### Vamos trabalhar agora para encontrar a Classificação do campeonado de pilotos, mas antes vamos trazer o ano referente ao resultado para facilitar a sumarização dos campeonatos

In [27]:
resultado_corrida = pd.merge(fato_resultado, dim_corrida[['ID_CORRIDA', 'ANO']], on='ID_CORRIDA', how='inner')
resultado_corrida.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  POS_FINAL  \
0             1          18          1              1            1          1   
1             2          18          2              2            5          2   
2             3          18          3              3            7          3   
3             4          18          4              4           11          4   
4             5          18          5              1            3          5   

   PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA    STATUS   ANO  
0              10                  2                   0  Finished  2008  
1               8                  3                   0  Finished  2008  
2               6                  5                   0  Finished  2008  
3               5                  7                   0  Finished  2008  
4               4                  1                   1  Finished  2008

In [28]:
# Dataframe sem valores nulos
resultado_corrida.isnull().sum()

ID_RESULTADO          0
ID_CORRIDA            0
ID_PILOTO             0
ID_CONSTRUTOR         0
POS_LARGADA           0
POS_FINAL             0
PONTOS_CORRIDA        0
RANK_VOLTA_RAPIDA     0
PONTO_VOLTA_RAPIDA    0
STATUS                0
ANO                   0
dtype: int64

In [86]:
# Aqui trazemos as informações da dimensão piloto, já com a informação do Ano
resultado_piloto = pd.merge(resultado_corrida, dim_piloto, on='ID_PILOTO', how='inner')
resultado_piloto.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  POS_FINAL  \
0             1          18          1              1            1          1   
1            27          19          1              1            9          5   
2            57          20          1              1            3         13   
3            69          21          1              1            5          3   
4            90          22          1              1            3          2   

   PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA    STATUS   ANO  \
0              10                  2                   0  Finished  2008   
1               4                  3                   0  Finished  2008   
2               0                 19                   0    +1 Lap  2008   
3               6                  3                   0  Finished  2008   
4               8                  2                   0  Finished  2008   

             PAIS NOME_CONSTRUTOR REF_PILOTO SIGLA_PILOTO NOME_PILOTO  \
0  United Kingdom         McLaren   hamilton          HAM       Lewis   
1  United Kingdom         McLaren   hamilton          HAM       Lewis   
2  United Kingdom         McLaren   hamilton          HAM       Lewis   
3  United Kingdom         McLaren   hamilton          HAM       Lewis   
4  United Kingdom         McLaren   hamilton          HAM       Lewis   

  SOBRENOME_PILOTO DATA_NASCIMENTO  
0         Hamilton      1985-01-07  
1         Hamilton      1985-01-07  
2         Hamilton      1985-01-07  
3         Hamilton      1985-01-07  
4         Hamilton      1985-01-07

In [87]:
# Aqui podemos ver os 30 pilotos com maiores pontuações nas temporadas
pilotos_30_maiores_pontos = resultado_piloto.groupby(['ANO', 'REF_PILOTO'])['PONTOS_CORRIDA'].agg('sum').sort_values(ascending=False)[:30]
pilotos_30_maiores_pontos

ANO   REF_PILOTO    
2019  hamilton          413
2018  hamilton          408
2013  vettel            397
2011  vettel            392
2016  rosberg           385
2014  hamilton          384
2015  hamilton          381
2016  hamilton          380
2017  hamilton          363
2020  hamilton          347
2019  bottas            326
2015  rosberg           322
2018  vettel            320
2014  rosberg           317
2017  vettel            317
      bottas            305
2012  vettel            281
2015  vettel            278
2019  max_verstappen    278
2012  alonso            278
2011  button            270
2019  leclerc           264
2011  webber            258
      alonso            257
2010  vettel            256
2016  ricciardo         256
2010  alonso            252
2018  raikkonen         251
      max_verstappen    249
      bottas            247
Name: PONTOS_CORRIDA, dtype: int64

In [88]:
# Aqui vemos quantas vezes cada piloto conseguiu as maiores pontuações na lista com os 30
result = pd.DataFrame(pilotos_30_maiores_pontos)
result.index.get_level_values('REF_PILOTO').value_counts()

vettel            7
hamilton          7
alonso            3
bottas            3
rosberg           3
max_verstappen    2
raikkonen         1
button            1
ricciardo         1
webber            1
leclerc           1
Name: REF_PILOTO, dtype: int64

In [89]:
# Vamos analisar o campeonato de pilotos 2020
resultado_piloto_2020 = resultado_piloto[resultado_piloto['ANO'] == 2020]
resultado_piloto_2020.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  \
250         24629        1031          1            131            5   
251         24646        1032          1            131            1   
252         24666        1033          1            131            1   
253         24686        1034          1            131            1   
254         24707        1035          1            131            2   

     POS_FINAL  PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA  \
250          4              12                  3                   0   
251          1              25                  3                   0   
252          1              26                  1                   1   
253          1              25                  3                   0   
254          2              19                  1                   1   

       STATUS   ANO            PAIS NOME_CONSTRUTOR REF_PILOTO SIGLA_PILOTO  \
250  Finished  2020  United Kingdom         McLaren   hamilton          HAM   
251  Finished  2020  United Kingdom         McLaren   hamilton          HAM   
252  Finished  2020  United Kingdom         McLaren   hamilton          HAM   
253  Finished  2020  United Kingdom         McLaren   hamilton          HAM   
254  Finished  2020  United Kingdom         McLaren   hamilton          HAM   

    NOME_PILOTO SOBRENOME_PILOTO DATA_NASCIMENTO  
250       Lewis         Hamilton      1985-01-07  
251       Lewis         Hamilton      1985-01-07  
252       Lewis         Hamilton      1985-01-07  
253       Lewis         Hamilton      1985-01-07  
254       Lewis         Hamilton      1985-01-07

In [90]:
# Classificação final de pilotos de 2020
resultado_piloto_2020.groupby(['REF_PILOTO'])['PONTOS_CORRIDA'].agg('sum').sort_values(ascending=False)


REF_PILOTO
hamilton             347
bottas               223
max_verstappen       214
perez                125
ricciardo            119
sainz                105
albon                105
leclerc               98
norris                97
gasly                 75
stroll                75
ocon                  62
vettel                33
kvyat                 32
hulkenberg            10
giovinazzi             4
raikkonen              4
russell                3
grosjean               2
kevin_magnussen        1
latifi                 0
pietro_fittipaldi      0
aitken                 0
Name: PONTOS_CORRIDA, dtype: int64

<h3 style="text-align:center;">Imagem com pesquisa no google do campeonato de pilotos da temporada de 2020</h3>

!["Campeonato de Pilotos de Fórmula 1 - Temporada de 2020"](./images/piloto_2020.jpg)

### Vamos analisar o campeonato de construtores

In [92]:
# Juntando os dataframes de resultado_corrida com a dimensao construtor
resultado_construtor = pd.merge(resultado_corrida, dim_construtor, on='ID_CONSTRUTOR', how='inner')
resultado_construtor.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  POS_FINAL  \
0             1          18          1              1            1          1   
1             5          18          5              1            3          5   
2            25          19          5              1            8          3   
3            27          19          1              1            9          5   
4            49          20          5              1            5          5   

   PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA    STATUS   ANO  \
0              10                  2                   0  Finished  2008   
1               4                  1                   1  Finished  2008   
2               6                  7                   0  Finished  2008   
3               4                  3                   0  Finished  2008   
4               4                  1                   1  Finished  2008   

             PAIS FABRICANTE_MOTOR NOME_CONSTRUTOR  
0  United Kingdom         Mercedes         McLaren  
1  United Kingdom         Mercedes         McLaren  
2  United Kingdom         Mercedes         McLaren  
3  United Kingdom         Mercedes         McLaren  
4  United Kingdom         Mercedes         McLaren

In [111]:
# As 50 maiores pontuações de construtores por temporada, guarde a informação da Mercedes em 2019
construtores_maiores_pontos = resultado_construtor.groupby(['ANO', 'NOME_CONSTRUTOR'])['PONTOS_CORRIDA'].agg('sum').sort_values(ascending=False)
construtores_maiores_pontos[:50]

ANO   NOME_CONSTRUTOR
2016  Mercedes           765
2019  Mercedes           739
2015  Mercedes           703
2014  Mercedes           701
2017  Mercedes           668
2018  Mercedes           655
2011  Red Bull           650
2013  Red Bull           596
2020  Mercedes           573
2018  Ferrari            571
2017  Ferrari            522
2019  Ferrari            504
2010  Red Bull           498
2011  McLaren            497
2016  Red Bull           468
2012  Red Bull           460
2010  McLaren            454
2015  Ferrari            428
2018  Red Bull           419
2019  Red Bull           417
2014  Red Bull           405
2012  Ferrari            400
2016  Ferrari            398
2010  Ferrari            396
2012  McLaren            378
2011  Ferrari            375
2017  Red Bull           368
2013  Mercedes           360
      Ferrari            354
2014  Williams           320
2020  Red Bull           319
2013  Lotus F1           315
2012  Lotus F1           303
2004  Ferrari        

In [99]:
# Aqui vemos quantas vezes cada construtor conseguiu as maiores pontuações 50 pontuações
construtor = pd.DataFrame(construtores_maiores_pontos[:50])
construtor.index.get_level_values('NOME_CONSTRUTOR').value_counts()

Ferrari         14
Red Bull        11
Mercedes         9
McLaren          7
Renault          3
Williams         2
Lotus F1         2
Racing Point     1
Force India      1
Name: NOME_CONSTRUTOR, dtype: int64

In [103]:
# Vamos analisar o campeonato de construtores de 2020
resultado_construtor_2020 = resultado_construtor[resultado_construtor['ANO'] == 2020]
resultado_construtor_2020.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  \
1709         24628        1031        846              1            3   
1710         24630        1031        832              1            8   
1711         24650        1032        846              1            9   
1712         24654        1032        832              1            3   
1713         24674        1033        832              1            9   

      POS_FINAL  PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA  \
1709          3              16                  1                   1   
1710          5              10                  5                   0   
1711          5              10                  5                   0   
1712          9               3                  1                   1   
1713          9               2                 15                   0   

        STATUS   ANO            PAIS FABRICANTE_MOTOR NOME_CONSTRUTOR  
1709  Finished  2020  United Kingdom         Mercedes         McLaren  
1710  Finished  2020  United Kingdom         Mercedes         McLaren  
1711  Finished  2020  United Kingdom         Mercedes         McLaren  
1712    +1 Lap  2020  United Kingdom         Mercedes         McLaren  
1713    +1 Lap  2020  United Kingdom         Mercedes         McLaren

In [104]:
# Classificação final do campeonato de construtores de 2020
resultado_construtor_2020.groupby(['NOME_CONSTRUTOR'])['PONTOS_CORRIDA'].agg('sum').sort_values(ascending=False)

NOME_CONSTRUTOR
Mercedes        573
Red Bull        319
Racing Point    210
McLaren         202
Renault         181
Ferrari         131
AlphaTauri      107
Alfa Romeo        8
Haas F1 Team      3
Williams          0
Name: PONTOS_CORRIDA, dtype: int64

<h3 style="text-align:center;">Imagem com pesquisa no google do campeonato de construtores da temporada de 2020</h3>

!["Campeonato de Construtor de Fórmula 1 - Temporada de 2020"](./images/construtor_2020.jpg)

<h2 style="color:red;">Mas pera aí!!! A Racing Point fez 210 pontos ou 195 pontos???</h2>

In [106]:
# Pegando os dados de resultado de 2020 da Racing Point, vemos que nesse ano foram 3 pilotos correndo pela Racing Point
resultado_construtor_2020[resultado_construtor_2020['NOME_CONSTRUTOR'] == 'Racing Point']

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  \
24956         24631        1031        815            211            6   
24957         24643        1031        840            211            9   
24958         24651        1032        815            211           17   
24959         24652        1032        840            211           12   
24960         24669        1033        840            211            3   
24961         24672        1033        815            211            4   
24962         24694        1034        840            211            6   
24963         24705        1034        807            211           13   
24964         24711        1035        840            211            6   
24965         24712        1035        807            211            3   
24966         24729        1036        840            211            5   
24967         24730        1036        815            211            4   
24968         24754        1037        840            211            9   
24969         24755        1037        815            211            8   
24970         24768        1038        840            211            8   
24971         24775        1038        815            211            4   
24972         24790        1039        815            211            7   
24973         24798        1039        840            211            6   
24974         24809        1040        815            211            4   
24975         24825        1040        840            211           12   
24976         24829        1041        815            211            9   
24977         24833        1041        807            211           20   
24978         24852        1042        815            211            5   
24979         24865        1042        840            211           12   
24980         24871        1043        815            211           11   
24981         24878        1043        840            211           15   
24982         24887        1044        815            211            3   
24983         24894        1044        840            211            1   
24984         24923        1045        815            211            5   
24985         24924        1045        840            211           13   
24986         24926        1046        815            211            5   
24987         24928        1046        840            211           10   
24988         24955        1047        840            211            8   
24989         24965        1047        815            211           19   

       POS_FINAL  PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA  \
24956          6               8                  6                   0   
24957         18               0                 18                   0   
24958          6               8                  4                   0   
24959          7               6                 10                   0   
24960          4              12                  3                   0   
24961          7               6                 10                   0   
24962          9               2                 13                   0   
24963         20               0                  0                   0   
24964          6               8                 15                   0   
24965          7               6                  6                   0   
24966          4              12                  8                   0   
24967          5              10                 16                   0   
24968          9               2                 13                   0   
24969         10               1                  9                   0   
24970          3              15                  3                   0   
24971         10               1                  8                   0   
24972          5              10                  5                   0   
24973         13               0                 12               

In [107]:
# Total de pontos de Hulkenberg
pontos_hulkenberg = resultado_construtor_2020[(resultado_construtor_2020['NOME_CONSTRUTOR'] == 'Racing Point') & (resultado_construtor_2020['ID_PILOTO'] == 807)]['PONTOS_CORRIDA'].sum()
pontos_hulkenberg

10

In [108]:
# Total de pontos de Sergio Perez
pontos_sergio_perez = resultado_construtor_2020[(resultado_construtor_2020['NOME_CONSTRUTOR'] == 'Racing Point') & (resultado_construtor_2020['ID_PILOTO'] == 815)]['PONTOS_CORRIDA'].sum()
pontos_sergio_perez

125

In [109]:
# Total de pontos de Lance Stroll
pontos_lance_stroll = resultado_construtor_2020[(resultado_construtor_2020['NOME_CONSTRUTOR'] == 'Racing Point') & (resultado_construtor_2020['ID_PILOTO'] == 840)]['PONTOS_CORRIDA'].sum()
pontos_lance_stroll

75

In [110]:
# Total de pontos da Racing Point para 2020
pontos_racing_point = pontos_hulkenberg + pontos_sergio_perez + pontos_lance_stroll
pontos_racing_point

210

<h3 style="text-align:center;">Encontramos a resposta após uma pesquisa, o motivo: Essa máquina Rosa!</h3>

!["Campeonato de Construtor de Fórmula 1 - Temporada de 2020"](./images/racing_car.jpg)


<p><strong>A Racing Point foi deduzida com 15 pontos no campeonato mundial e multada em € 400.000 depois que os comissários da FIA confirmaram o protesto da Renault sobre a legalidade do design do carro RP20 da equipe.</strong></p>

<p>Após os Grandes Prêmios Styrian, Húngaro e Britânico, a Renault apresentou um protesto com os comissários, alegando que os dutos de freio da Racing Point haviam sido copiados diretamente da Mercedes 2019.</p>

<p>E na sexta-feira em Silverstone, os comissários deram seu veredicto, apoiando o protesto da Renault. A Racing Point foi multada em € 200.000 por carro. Eles também foram deduzidos 7,5 pontos do campeonato mundial por carro - embora Perez e Stroll mantenham os pontos do campeonato de seus pilotos. Protestos pelo Grande Prêmio da Hungria e Da Inglaterra também foram confirmados, com a Racing Point repreendida por usar seu carro em ambos os eventos.</p>

<p>Em um documento de 14 páginas emitido em Silverstone, os comissários da FIA descreveram em detalhes por que o protesto da Renault havia sido confirmado. O cerne da questão se concentrou em quando as novas regras em torno das peças listadas foram aplicadas, a Racing Point argumentando que os novos regulamentos esportivos só entraram em vigor um dia antes da FP1 na Áustria. Os comissários rejeitaram essa noção, e disseram que, embora o atual design do duto de freio dianteiro da equipe fosse aceitável, dado que era uma evolução do projeto da equipe para 2019 – concebida usando desenhos CAD dos dutos de freio do carro W10 da Mercedes, quando tal prática era permitida pelo regulamento – os dutos de freio traseiros da equipe em 2020 devem ser considerados projetos da Mercedes. Isso porque, embora tenham sido projetados usando os mesmos materiais, a Racing Point não estava refinando um componente que já havia sido incorporado ao DNA do RP19. Em vez disso, eles estavam - aos olhos dos administradores - introduzindo um componente completamente novo para o RP20, que ele sabia que agora era classificado como uma parte listada.</p>

<p>Apesar da decisão, a Racing Point continuará usando os dutos de freio daqui para frente, começando com o Grande Prêmio do 70º Aniversário deste fim de semana. Isso porque os administradores reconheceram que não era realista esperar que a Racing Point "reescaria ou re‐projetasse os BDs de uma maneira que efetivamente exigiria que eles 'desaprendessem' o que já sabem". A equipe está atualmente digerindo os resultados da investigação antes de decidir se deve ou não recorrer da decisão, em consonância com o artigo 15 do Código Esportivo Internacional. Falando ao F1.com logo após o diretor da Racing Point Team, Otmar Szafnauer, dizer que sua reação inicial é que a decisão foi "injusta".</p>

Racing Point drivers scored 210 points, but the constructor was deducted 15 points after a protest from Renault was upheld regarding the legality of their car. <a href="https://www.formula1.com/en/latest/article.breaking-racing-point-deducted-15-points-and-fined-heavily-as-renault.7j9tsLSAm7rXnWHxH1Uy9c.html" target="_blank">Link para Site da F1</a>



<h3 style="text-align:center;">Veja algumas imagens comparando a Racing Point 2020 com a Mercedes 2019</h3>

!["Comparativo Racing Point 2020 vs Mercedes 2019"](./images/racing_car1.jpg)

!["Comparativo Racing Point 2020 vs Mercedes 2019"](./images/racing_car2.jpg)

!["Comparativo Racing Point 2020 vs Mercedes 2019"](./images/racing_car3.jpg)

!["Comparativo Racing Point 2020 vs Mercedes 2019"](./images/racing_car4.jpg)

In [112]:
# Vamos analisar quais os pilotos com mais pódios
resultado_piloto_podios = resultado_piloto[resultado_piloto['POS_FINAL'].isin([1, 2, 3])]
resultado_piloto_podios.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  POS_FINAL  \
0             1          18          1              1            1          1   
3            69          21          1              1            5          3   
4            90          22          1              1            3          2   
5           109          23          1              1            3          1   
8           169          26          1              1            4          1   

   PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA    STATUS   ANO  \
0              10                  2                   0  Finished  2008   
3               6                  3                   0  Finished  2008   
4               8                  2                   0  Finished  2008   
5              10                  6                   0  Finished  2008   
8              10                  3                   0  Finished  2008   

             PAIS NOME_CONSTRUTOR REF_PILOTO SIGLA_PILOTO NOME_PILOTO  \
0  United Kingdom         McLaren   hamilton          HAM       Lewis   
3  United Kingdom         McLaren   hamilton          HAM       Lewis   
4  United Kingdom         McLaren   hamilton          HAM       Lewis   
5  United Kingdom         McLaren   hamilton          HAM       Lewis   
8  United Kingdom         McLaren   hamilton          HAM       Lewis   

  SOBRENOME_PILOTO DATA_NASCIMENTO  
0         Hamilton      1985-01-07  
3         Hamilton      1985-01-07  
4         Hamilton      1985-01-07  
5         Hamilton      1985-01-07  
8         Hamilton      1985-01-07

In [114]:
# Resultado com os 30 pilotos com mais pódios de toda história
resultado_piloto_podios.groupby(['REF_PILOTO'])['POS_FINAL'].agg('count').sort_values(ascending=False)[:30]

REF_PILOTO
hamilton              169
michael_schumacher    155
vettel                121
prost                 106
raikkonen             103
alonso                 97
senna                  80
barrichello            68
coulthard              62
piquet                 60
mansell                59
bottas                 59
rosberg                57
lauda                  54
hakkinen               51
button                 50
berger                 48
max_verstappen         46
reutemann              45
stewart                43
webber                 42
damon_hill             42
massa                  41
patrese                37
hill                   36
emerson_fittipaldi     35
fangio                 35
scheckter              33
laffite                33
hulme                  33
Name: POS_FINAL, dtype: int64

In [115]:
# Vamos analisar quais os pilotos com mais poles
resultado_piloto_poles = resultado_piloto[resultado_piloto['POS_LARGADA'] == 1]
resultado_piloto_poles.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  \
0              1          18          1              1            1   
6            147          24          1              1            1   
9            189          27          1              1            1   
10           213          28          1              1            1   
12           251          30          1              1            1   

    POS_FINAL  PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA  \
0           1              10                  2                   0   
6          19               0                  3                   0   
9           1              10                  2                   0   
10          5               4                  3                   0   
12          3               6                  2                   0   

       STATUS   ANO            PAIS NOME_CONSTRUTOR REF_PILOTO SIGLA_PILOTO  \
0    Finished  2008  United Kingdom         McLaren   hamilton          HAM   
6   Collision  2008  United Kingdom         McLaren   hamilton          HAM   
9    Finished  2008  United Kingdom         McLaren   hamilton          HAM   
10   Finished  2008  United Kingdom         McLaren   hamilton          HAM   
12   Finished  2008  United Kingdom         McLaren   hamilton          HAM   

   NOME_PILOTO SOBRENOME_PILOTO DATA_NASCIMENTO  
0        Lewis         Hamilton      1985-01-07  
6        Lewis         Hamilton      1985-01-07  
9        Lewis         Hamilton      1985-01-07  
10       Lewis         Hamilton      1985-01-07  
12       Lewis         Hamilton      1985-01-07

In [116]:
# Resultado com os 30 pilotos com mais poles de toda história
resultado_piloto_poles.groupby(['REF_PILOTO'])['POS_LARGADA'].agg('count').sort_values(ascending=False)[:30]

REF_PILOTO
hamilton              100
michael_schumacher     68
senna                  65
vettel                 57
clark                  34
prost                  33
mansell                32
rosberg                30
fangio                 29
hakkinen               26
piquet                 24
lauda                  24
alonso                 22
damon_hill             20
mario_andretti         18
raikkonen              18
arnoux                 18
moss                   17
bottas                 17
stewart                17
massa                  16
peterson               14
barrichello            14
ascari                 14
hunt                   14
montoya                13
jack_brabham           13
webber                 13
ickx                   13
hill                   13
Name: POS_LARGADA, dtype: int64

In [117]:
# Vamos analisar quais os pilotos com mais voltas rápidas
resultado_piloto_volta_rapida = resultado_piloto[resultado_piloto['RANK_VOLTA_RAPIDA'] == 1]
resultado_piloto_volta_rapida.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  \
16           329          34          1              1            1   
19           392          37          1              1            4   
32           677          50          1              1            1   
55         20396         340          1              1            6   
56         20432         341          1              1            3   

    POS_FINAL  PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA  \
16          1              10                  1                   1   
19          2               8                  1                   1   
32          1              10                  1                   1   
55          2              18                  1                   1   
56         14               0                  1                   0   

      STATUS   ANO            PAIS NOME_CONSTRUTOR REF_PILOTO SIGLA_PILOTO  \
16  Finished  2008  United Kingdom         McLaren   hamilton          HAM   
19  Finished  2007  United Kingdom         McLaren   hamilton          HAM   
32  Finished  2007  United Kingdom         McLaren   hamilton          HAM   
55  Finished  2010  United Kingdom         McLaren   hamilton          HAM   
56  Puncture  2010  United Kingdom         McLaren   hamilton          HAM   

   NOME_PILOTO SOBRENOME_PILOTO DATA_NASCIMENTO  
16       Lewis         Hamilton      1985-01-07  
19       Lewis         Hamilton      1985-01-07  
32       Lewis         Hamilton      1985-01-07  
55       Lewis         Hamilton      1985-01-07  
56       Lewis         Hamilton      1985-01-07

In [118]:
# Resultado com os 30 pilotos com mais voltas rápidas de toda história
resultado_piloto_volta_rapida.groupby(['REF_PILOTO'])['RANK_VOLTA_RAPIDA'].agg('count').sort_values(ascending=False)[:30]

REF_PILOTO
hamilton              54
raikkonen             42
vettel                38
alonso                22
michael_schumacher    21
rosberg               20
webber                19
bottas                17
massa                 15
ricciardo             15
max_verstappen        11
button                 8
barrichello            6
leclerc                4
perez                  4
montoya                3
norris                 2
gasly                  2
kovalainen             2
kevin_magnussen        2
hulkenberg             2
heidfeld               2
gutierrez              1
fisichella             1
glock                  1
bruno_senna            1
grosjean               1
rosa                   1
trulli                 1
kubica                 1
Name: RANK_VOLTA_RAPIDA, dtype: int64

In [119]:
# Vamos analisar quais os pilotos com mais vitórias de toda história
resultado_piloto_vitorias = resultado_piloto[resultado_piloto['POS_FINAL'] == 1]
resultado_piloto_vitorias.head()

ID_RESULTADO  ID_CORRIDA  ID_PILOTO  ID_CONSTRUTOR  POS_LARGADA  \
0              1          18          1              1            1   
5            109          23          1              1            3   
8            169          26          1              1            4   
9            189          27          1              1            1   
16           329          34          1              1            1   

    POS_FINAL  PONTOS_CORRIDA  RANK_VOLTA_RAPIDA  PONTO_VOLTA_RAPIDA  \
0           1              10                  2                   0   
5           1              10                  6                   0   
8           1              10                  3                   0   
9           1              10                  2                   0   
16          1              10                  1                   1   

      STATUS   ANO            PAIS NOME_CONSTRUTOR REF_PILOTO SIGLA_PILOTO  \
0   Finished  2008  United Kingdom         McLaren   hamilton          HAM   
5   Finished  2008  United Kingdom         McLaren   hamilton          HAM   
8   Finished  2008  United Kingdom         McLaren   hamilton          HAM   
9   Finished  2008  United Kingdom         McLaren   hamilton          HAM   
16  Finished  2008  United Kingdom         McLaren   hamilton          HAM   

   NOME_PILOTO SOBRENOME_PILOTO DATA_NASCIMENTO  
0        Lewis         Hamilton      1985-01-07  
5        Lewis         Hamilton      1985-01-07  
8        Lewis         Hamilton      1985-01-07  
9        Lewis         Hamilton      1985-01-07  
16       Lewis         Hamilton      1985-01-07

In [120]:
# Resultado com os 30 pilotos com mais vitórias de toda história
piloto_vitorias = resultado_piloto_vitorias.groupby(['REF_PILOTO'])['POS_FINAL'].agg('count').sort_values(ascending=False)
piloto_vitorias[:30]

REF_PILOTO
hamilton              98
michael_schumacher    91
vettel                53
prost                 51
senna                 41
alonso                32
mansell               31
stewart               27
clark                 25
lauda                 25
fangio                24
rosberg               23
piquet                23
damon_hill            22
raikkonen             21
hakkinen              20
moss                  16
button                15
jack_brabham          14
emerson_fittipaldi    14
hill                  14
coulthard             13
ascari                13
jones                 12
mario_andretti        12
reutemann             12
villeneuve            11
barrichello           11
max_verstappen        11
massa                 11
Name: POS_FINAL, dtype: int64

In [121]:
# Vamos analisar quais os pilotos com mais corridas de toda história
piloto_corridas = resultado_piloto.groupby(['REF_PILOTO'])['POS_LARGADA'].agg('count').sort_values(ascending=False)
piloto_corridas[:30]

REF_PILOTO
raikkonen             336
barrichello           326
alonso                318
button                309
michael_schumacher    308
massa                 271
hamilton              270
vettel                262
patrese               257
trulli                256
coulthard             247
fisichella            231
webber                217
alboreto              215
cesaris               214
berger                210
piquet                207
rosberg               206
prost                 202
alesi                 202
perez                 197
ricciardo             192
mansell               192
heidfeld              184
hulkenberg            182
grosjean              181
ralf_schumacher       180
laffite               180
hill                  179
lauda                 174
Name: POS_LARGADA, dtype: int64

In [122]:
# Com as informações de total de vitórias e total de corridas, vamos analisar a performance com o % de vitórias por corrida
frames = [piloto_vitorias, piloto_corridas] 
vitorias_corrida = pd.concat(frames, axis=1)
vitorias_corrida.head()

POS_FINAL  POS_LARGADA
hamilton                 98.0          270
michael_schumacher       91.0          308
vettel                   53.0          262
prost                    51.0          202
senna                    41.0          162

In [123]:
# Renomeando as colunas corretamente
vitorias_corrida.rename(columns={"POS_FINAL": "TOTAL_VITORIAS", "POS_LARGADA": "TOTAL_CORRIDAS"}, inplace=True)

In [124]:
# Calculando o % de desempenho dos pilotos que venceram vs corridas que participaram
vitorias_corrida['%_DESEMPENHO'] = vitorias_corrida['TOTAL_VITORIAS'] / vitorias_corrida['TOTAL_CORRIDAS'] * 100
vitorias_corrida.head()

TOTAL_VITORIAS  TOTAL_CORRIDAS  %_DESEMPENHO
hamilton                      98.0             270     36.296296
michael_schumacher            91.0             308     29.545455
vettel                        53.0             262     20.229008
prost                         51.0             202     25.247525
senna                         41.0             162     25.308642

In [126]:
# Resultado com os pilotos com melhor performance de vitórias por corrida, mas quem é Wallard e Vikovich?
vitorias_corrida['%_DESEMPENHO'].sort_values(ascending=False)[:20]

wallard               50.000000
fangio                41.379310
vukovich              40.000000
hamilton              36.296296
ascari                36.111111
clark                 34.246575
michael_schumacher    29.545455
stewart               27.000000
senna                 25.308642
prost                 25.247525
moss                  21.917808
vettel                20.229008
sweikert              20.000000
damon_hill            18.032787
flaherty              16.666667
mansell               16.145833
brooks                14.634146
lauda                 14.367816
farina                13.513514
fagioli               12.500000
Name: %_DESEMPENHO, dtype: float64

<h3>Identificamos que esses pilotos tem valores inteiros, quantas corridas eles disputaram?</h3>

In [127]:
# Quantidade de corridas de wallard
piloto_corridas.loc['wallard']

2

In [129]:
# Quantidade de vitórias de wallard
piloto_vitorias.loc['wallard']

1

In [130]:
# Quantidade de corridas de vukovich
piloto_corridas.loc['vukovich']

5

In [131]:
# Quantidade de vitórias de vukovich
piloto_vitorias.loc['vukovich']

2

<h3>Para não poluir nosso resultado, vamos filtrar apenas os pilotos que correram ao menos 32 corridas (+- 2 temporadas) e venceram ao menos 1</h3>

In [132]:
vitorias_32_corrida_ou_mais = vitorias_corrida[(vitorias_corrida['TOTAL_CORRIDAS'] >= 32) & (vitorias_corrida['TOTAL_VITORIAS'] >= 1)]
vitorias_32_corrida_ou_mais.head(10)

TOTAL_VITORIAS  TOTAL_CORRIDAS  %_DESEMPENHO
hamilton                      98.0             270     36.296296
michael_schumacher            91.0             308     29.545455
vettel                        53.0             262     20.229008
prost                         51.0             202     25.247525
senna                         41.0             162     25.308642
alonso                        32.0             318     10.062893
mansell                       31.0             192     16.145833
stewart                       27.0             100     27.000000
clark                         25.0              73     34.246575
lauda                         25.0             174     14.367816

In [133]:
# Resultado final com os pilotos com melhor % de desempenho de vitoria por corrida disputada
vitorias_32_corrida_ou_mais['%_DESEMPENHO'].sort_values(ascending=False)[:20]

fangio                41.379310
hamilton              36.296296
ascari                36.111111
clark                 34.246575
michael_schumacher    29.545455
stewart               27.000000
senna                 25.308642
prost                 25.247525
moss                  21.917808
vettel                20.229008
damon_hill            18.032787
mansell               16.145833
brooks                14.634146
lauda                 14.367816
farina                13.513514
hakkinen              12.121212
rosberg               11.165049
piquet                11.111111
jack_brabham          10.852713
hunt                  10.752688
Name: %_DESEMPENHO, dtype: float64

<h1>In Memorian... Fangio</h1>

<p>Juan Manuel Fangio (24 de junho de 1911 - 17 de julho de 1995), apelidado de El Chueco ou El Maestro, foi um piloto de automobilismo argentino. Ele dominou a primeira década de corridas de Fórmula 1, vencendo o Campeonato Mundial de Pilotos cinco vezes.

Ele venceu o Campeonato Mundial de Pilotos cinco vezes — um recorde que permaneceu por 46 anos até ser batido por Michael Schumacher— com quatro equipes diferentes (Alfa Romeo, Ferrari, Mercedes-Benz e Maserati), um feito que não se repetiu.
</p>

!["Juan Manuel Fangio"](./images/fangio.jpg)

<h1 style="text-align: center">Grato!</h1>

!["Lewis Hamilton"](./images/hamilton.jpg)